tests/common/gnmi_setup.py

```diff
@@ -0,0 +1,347 @@
+    dut_command = "docker exec %s cat /root/gnmi.log" % (env.gnmi_container)
+    res = duthost.shell(dut_command, module_ignore_errors=True)
+    logger.info("GNMI log: " + res['stdout'])
+    return res['stdout']
+
+
+def dump_system_status(duthost):
+    env = GNMIEnvironment(duthost, GNMIEnvironment.GNMI_MODE)
+    dut_command = "docker exec %s ps -efwww" % (env.gnmi_container)
+    res = duthost.shell(dut_command, module_ignore_errors=True)
+    logger.info("GNMI process: " + res['stdout'])
+    dut_command = "docker exec %s date" % (env.gnmi_container)
+    res = duthost.shell(dut_command, module_ignore_errors=True)
+    logger.info("System time: " + res['stdout'] + res['stderr'])
+
+
+def check_gnmi_status(duthost):
+    env = GNMIEnvironment(duthost, GNMIEnvironment.GNMI_MODE)
+    dut_command = "docker exec %s supervisorctl status %s" % (env.gnmi_container, env.gnmi_program)
+    output = duthost.shell(dut_command, module_ignore_errors=True)
+    return "RUNNING" in output['stdout']
+
+
+def apply_cert_config(duthost):
+    env = GNMIEnvironment(duthost, GNMIEnvironment.GNMI_MODE)
+    # Get subtype
+    cfg_facts = duthost.config_facts(host=duthost.hostname, source="running")['ansible_facts']
+    metadata = cfg_facts["DEVICE_METADATA"]["localhost"]
+    subtype = metadata.get('subtype', None)
+    # Stop all running program
+    dut_command = "docker exec %s supervisorctl status" % (env.gnmi_container)
+    output = duthost.shell(dut_command, module_ignore_errors=True)
+    for line in output['stdout_lines']:
+        res = line.split()
+        if len(res) < 3:
+            continue
+        program = res[0]
+        status = res[1]
+        if status == "RUNNING":
+            dut_command = "docker exec %s supervisorctl stop %s" % (env.gnmi_container, program)
+            duthost.shell(dut_command, module_ignore_errors=True)
+    dut_command = "docker exec %s pkill %s" % (env.gnmi_container, env.gnmi_process)
+    duthost.shell(dut_command, module_ignore_errors=True)
+    dut_command = "docker exec %s bash -c " % env.gnmi_container
+    dut_command += "\"/usr/bin/nohup /usr/sbin/%s -logtostderr --port %s " % (env.gnmi_process, env.gnmi_port)
+    dut_command += "--server_crt /etc/sonic/telemetry/gnmiserver.crt --server_key /etc/sonic/telemetry/gnmiserver.key "
+    dut_command += "--config_table_name GNMI_CLIENT_CERT "
+    dut_command += "--client_auth cert "
+    dut_command += "--enable_crl=true "
+    if subtype == 'SmartSwitch':
+        dut_command += "--zmq_address=tcp://127.0.0.1:8100 "
+    dut_command += "--ca_crt /etc/sonic/telemetry/gnmiCA.pem -gnmi_native_write=true -v=10 >/root/gnmi.log 2>&1 &\""
+    duthost.shell(dut_command)
+
+    # Setup gnmi client cert common name
+    add_gnmi_client_common_name(duthost, "test.client.gnmi.sonic")
+    add_gnmi_client_common_name(duthost, "test.client.revoked.gnmi.sonic")
+
+    time.sleep(GNMI_SERVER_START_WAIT_TIME)
```

Comment:
>GNMI_SERVER_START_WAIT_TIME [](http://example.com/codeflow?start=15&length=27)

you may reduce the wait time: as long as the port is connectable, you just stop waiting.

Reply:
I have refactored it a bit. I plan to refactor this module after this PR is committed as much of it's a copy from `tests/gnmi/helper.py`. I intend to refactor this and then gnmi tests in a separate PR.